Okta.Xamarin/Okta.Xamarin/OidcClient.cs

```diff
@@ -13,279 +13,362 @@
-		public async Task<StateManager> AuthenticateAsync(string sessionToken)
+    /// <summary>
+    /// A client for logging into Okta via Oidc
+    /// </summary>
+    public partial class OidcClient : IOidcClient
+    {
+        /// <summary>
+        /// The configuration for this Client.
+        /// </summary>
+        public IOktaConfig Config { get; set; }
+
+        /// <summary>
+        /// Maintains a list of all currently active Clients, by state.  This is used after the intent/universal link callback from login to continue the state machine.
+        /// </summary>
+        internal static Dictionary<string, IOidcClient> authenticatorsByState = new Dictionary<string, IOidcClient>();
+
+		internal static Dictionary<string, IOidcClient> loggingOutClientsByState = new Dictionary<string, IOidcClient>();
+
+        /// <summary>
+        /// The <see cref="HttpClient"/> for use in getting an auth token.  Microsoft guidance specifies that this should be reused for performance reasons.
+        /// </summary>
+        private HttpClient client = new HttpClient();
+
+        /// <summary>
+        /// A <see cref="OktaConfigValidator"/> used to validate any configuration used by Clients
+        /// </summary>
+        private static readonly OktaConfigValidator<IOktaConfig> validator = new OktaConfigValidator<IOktaConfig>();
+
+        /// <summary>
+        /// Start the authorization flow.  This is an async method and should be awaited.
+        /// </summary>
+        /// <returns>In case of successful authorization, this Task will return a valid <see cref="OktaState"/>.  Clients are responsible for further storage and maintenance of the manager.</returns>
+        public Task<OktaState> SignInWithBrowserAsync()
+        {
+            validator.Validate(Config);
+
+            currentTask = new TaskCompletionSource<OktaState>();
+            GenerateStateCodeVerifierAndChallenge();
+            authenticatorsByState.Add(State, this);
+            this.LaunchBrowser(this.GenerateAuthorizeUrl());
+
+            return currentTask.Task;
+        }
+
+        /// <summary>
+        /// This method will end the user's Okta session in the browser.  
+        /// </summary>
+        /// <param name="stateManager">The state manager associated with the login that you wish to log out</param>
+        /// <returns>Task which tracks the progress of the logout</returns>
+        public Task<OktaState> SignOutOfOktaAsync(OktaState stateManager)
+        {
+            validator.Validate(this.Config);
+            this.currentTask = new TaskCompletionSource<OktaState>();
+            if (!stateManager.IsAuthenticated)
+            {
+				this.currentTask.SetResult(stateManager);
+				return this.currentTask.Task;
+			}
+			this.GenerateStateCodeVerifierAndChallenge();
```

Comment:
Question for my own education. Do we also need to challenge when logging out when using PKCE?

Reply:
I don't think it is strictly required, just a good idea; however I'm not certain either way.